File: scala/difference-of-squares/src/main/scala/Squares.scala
object Squares {
  def squareOfSums(num: Int): Int = squared((0 to num).reduce(_ + _))
  def sumOfSquares(num: Int): Int = (0 to num).reduce(_ + squared(_))
  def difference(num: Int): Int = squareOfSums(num) - sumOfSquares(num)
  private def squared(num: Int): Int = num * num
}

File: haskell/difference-of-squares/src/Squares.hs
module Squares (difference, squareOfSums, sumOfSquares) where

difference :: Integral a => a -> a
difference n = squareOfSums n - sumOfSquares n

squareOfSums :: Integral a => a -> a
squareOfSums n =
  let total = sum [1..n]
  in total ^ 2

sumOfSquares :: Integral a => a -> a
sumOfSquares n =
  let squares = map (^2) [1..n]
  in sum squares
